# Download and Clean Dataset

Let's start by importing the pandas and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<h5>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different 
ingredients that were used to make them. Ingredients include: </h5>

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate

Let's read the dataset into a pandas dataframe.

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30


So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 
162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine 
aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa

<h3>Let's check how many data points we have.</h3>

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to 
overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

<h5>Split data into predictors and target</h5>


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:

concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:

predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [9]:

n_cols = predictors.shape[1] # number of predictors
n_cols

8

<h2>Import Keras</h2>


<h5>Let's go ahead and import the Keras library</h5>

In [10]:
import keras
from keras.models import Sequential
from tensorflow.keras.layers import Dense

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [11]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. 
It uses the adam optimizer and the mean squared error as the loss function.


Let's import scikit-learn in order to randomly split the data into a training and test sets

In [12]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [13]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

<h2>Train and Test the Network</h2>


Let's call the function now to create our model.

In [14]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.

In [15]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
23/23 [==============================] - 0s 688us/step - loss: 68989.7891
Epoch 2/50
23/23 [==============================] - 0s 730us/step - loss: 14597.9512
Epoch 3/50
23/23 [==============================] - 0s 735us/step - loss: 3668.0369
Epoch 4/50
23/23 [==============================] - 0s 783us/step - loss: 3200.7283
Epoch 5/50
23/23 [==============================] - 0s 838us/step - loss: 3094.9917
Epoch 6/50
23/23 [==============================] - 0s 781us/step - loss: 3018.4016
Epoch 7/50
23/23 [==============================] - 0s 780us/step - loss: 2943.4941
Epoch 8/50
23/23 [==============================] - 0s 842us/step - loss: 2862.0513
Epoch 9/50
23/23 [==============================] - 0s 824us/step - loss: 2785.2024
Epoch 10/50
23/23 [==============================] - ETA: 0s - loss: 2624.25 - 0s 804us/step - loss: 2713.6304
Epoch 11/50
23/23 [==============================] - 0s 848us/step - loss: 2646.5469
Epoch 12/50
23/23 [===========================

Next we need to evaluate the model on the test data.

In [16]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 599us/step - loss: 665.0792


665.0791625976562

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [17]:
from sklearn.metrics import mean_squared_error

In [18]:

mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

665.0791519278149 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [19]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 146.11367797851562
MSE 2: 132.71743774414062
MSE 3: 110.65909576416016
MSE 4: 130.17822265625
MSE 5: 148.4514617919922
MSE 6: 109.6833267211914
MSE 7: 130.99649047851562
MSE 8: 101.75447845458984
MSE 9: 118.67665100097656
MSE 10: 113.378173828125
MSE 11: 103.18502044677734
MSE 12: 101.40575408935547
MSE 13: 115.37207794189453
MSE 14: 118.96246337890625
MSE 15: 108.52578735351562
MSE 16: 121.33487701416016
MSE 17: 106.4878921508789
MSE 18: 106.13787841796875
MSE 19: 99.8309326171875
MSE 20: 117.38323974609375
MSE 21: 97.53101348876953
MSE 22: 110.05721282958984
MSE 23: 62.00568389892578
MSE 24: 54.47751235961914
MSE 25: 55.50408172607422
MSE 26: 62.156002044677734
MSE 27: 53.71994400024414
MSE 28: 46.39132308959961
MSE 29: 57.631256103515625
MSE 30: 51.286354064941406
MSE 31: 53.62424087524414
MSE 32: 43.605682373046875
MSE 33: 66.76485443115234
MSE 34: 49.43687057495117
MSE 35: 57.23036193847656
MSE 36: 59.30756378173828
MSE 37: 53.306854248046875
MSE 38: 56.01213455200195
MSE 3